File: sdk/typescript/src/client/http-client.ts
/**
 * @fileoverview Type-safe HTTP client with automatic retry and request signing
 * @author SwappiQ Protocol
 * @description Production-ready HTTP client with comprehensive error handling, retry logic, and security features
 */

import { EventEmitter } from 'events';
import {
  ApiResponse,
  ApiError,
  AuthCredentials,
  SignedRequest,
  RetryConfig,
  RateLimitConfig,
  SDKConfig
} from '../types/api.js';
import { RequestSigner } from '../utils/request-signer.js';
import { RateLimiter } from '../utils/rate-limiter.js';
import { 
  performanceMonitor, 
  CircuitBreaker, 
  LRUCache, 
  ObjectPool,
  trackPerformance 
} from '../utils/performance-monitor.js';

export interface RequestOptions {
  method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
  path: string;
  body?: any;
  headers?: Record<string, string>;
  auth?: boolean;
  timeout?: number;
  retries?: number;
  priority?: 'low' | 'normal' | 'high';
}

export interface RequestMetrics {
  requestId: string;
  method: string;
  path: string;
  startTime: number;
  endTime: number;
  duration: number;
  status: number;
  success: boolean;
  retryCount: number;
  fromCache: boolean;
}

export interface CachedResponse {
  data: ApiResponse<any>;
  timestamp: number;
}

export interface RequestContext {
  headers: Record<string, string>;
  body: any;
  metadata: Record<string, any>;
}

/**
 * Production-grade HTTP client with enterprise features
 */
export class HttpClient extends EventEmitter {
  private readonly baseUrl: string;
  private readonly auth?: AuthCredentials;
  private readonly retryConfig: RetryConfig;
  private readonly rateLimitConfig?: RateLimitConfig;
  private readonly timeout: number;
  private readonly debug: boolean;

  private readonly requestSigner: RequestSigner;
  private readonly rateLimiter?: RateLimiter;
  private readonly responseCache: LRUCache<string, CachedResponse>;
  private readonly activeRequests: Map<string, Promise<any>>;
  private readonly circuitBreaker: CircuitBreaker;
  private readonly requestPool: ObjectPool<RequestContext>;

  private requestCounter = 0;
  private metrics: RequestMetrics[] = [];

  constructor(config: SDKConfig) {
    super();

    this.baseUrl = config.apiUrl.replace(/\/$/, '');
    this.auth = config.auth;
    this.retryConfig = config.retryConfig;
    this.rateLimitConfig = config.rateLimitConfig;
    this.timeout = config.timeout;
    this.debug = config.debug || false;

    this.requestSigner = new RequestSigner(config.auth);
    this.rateLimiter = config.rateLimitConfig ? new RateLimiter(config.rateLimitConfig) : undefined;
    this.responseCache = new LRUCache(1000, 300000); // 1000 entries, 5 min TTL
    this.activeRequests = new Map();
    this.circuitBreaker = new CircuitBreaker(5, 60000, 2); // 5 failures, 1 min recovery
    this.requestPool = new ObjectPool(
      () => ({ headers: {}, body: null, metadata: {} }),
      (obj) => { obj.headers = {}; obj.body = null; obj.metadata = {}; },
      100 // Pool size
    );

    // Cleanup old metrics and cache entries
    setInterval(() => this.cleanup(), 300000); // Every 5 minutes
  }

  /**
   * Make a type-safe API request with automatic retry and rate limiting
   */
  @trackPerformance('http.request')
  async request<T = any>(options: RequestOptions): Promise<ApiResponse<T>> {
    const timer = performanceMonitor.startTimer('http.request');
    const requestId = this.generateRequestId();
    const startTime = Date.now();

    // Check circuit breaker
    if (this.circuitBreaker.isOpen()) {
      throw new Error('Circuit breaker is open - service temporarily unavailable');
    }

    try {
      // Rate limiting check
      if (this.rateLimiter) {
        await this.rateLimiter.acquire(options.priority);
      }

      // Check for duplicate requests
      const duplicateKey = this.getDuplicateKey(options);
      if (this.activeRequests.has(duplicateKey)) {
        if (this.debug) {
          console.log(`[${requestId}] Waiting for duplicate request: ${duplicateKey}`);
        }
        const result = await this.activeRequests.get(duplicateKey) as Promise<ApiResponse<T>>;
        timer.endWithMetadata({ fromCache: false, duplicate: true });
        return result;
      }

      // Check cache for GET requests
      if (options.method === 'GET') {
        const cached = this.responseCache.get(options.path);
        if (cached) {
          this.recordMetrics({
            requestId,
            method: options.method,
            path: options.path,
            startTime,
            endTime: Date.now(),
            duration: 0,
            status: 200,
            success: true,
            retryCount: 0,
            fromCache: true
          });
          timer.endWithMetadata({ fromCache: true, success: true });
          this.circuitBreaker.recordSuccess();
          return cached.data as ApiResponse<T>;
        }
      }

      // Create the request promise
      const requestPromise = this.executeRequest<T>(options, requestId, startTime);
      this.activeRequests.set(duplicateKey, requestPromise);

      try {
        const response = await requestPromise;
        
        // Cache successful GET responses
        if (options.method === 'GET' && response.success) {
          this.cacheResponse(options.path, response);
        }

        return response;
      } finally {
        this.activeRequests.delete(duplicateKey);
      }

    } catch (error) {
      this.recordMetrics({
        requestId,
        method: options.method,
        path: options.path,
        startTime,
        endTime: Date.now(),
        duration: Date.now() - startTime,
        status: 0,
        success: false,
        retryCount: 0,
        fromCache: false
      });

      throw this.createApiError(error, requestId);
    }
  }

  /**
   * Execute request with retry logic
   */
  private async executeRequest<T>(
    options: RequestOptions,
    requestId: string,
    startTime: number
  ): Promise<ApiResponse<T>> {
    let lastError: Error;
    let retryCount = 0;
    const maxRetries = options.retries ?? this.retryConfig.maxAttempts;

    while (retryCount <= maxRetries) {
      try {
        const response = await this.makeHttpRequest<T>(options, requestId);
        
        this.recordMetrics({
          requestId,
          method: options.method,
          path: options.path,
          startTime,
          endTime: Date.now(),
          duration: Date.now() - startTime,
          status: 200, // Assuming success if no error thrown
          success: response.success,
          retryCount,
          fromCache: false
        });

        return response;

      } catch (error) {
        lastError = error as Error;
        
        if (retryCount === maxRetries || !this.isRetryableError(error)) {
          break;
        }

        retryCount++;
        const delay = this.calculateRetryDelay(retryCount);
        
        if (this.debug) {
          console.warn(`[${requestId}] Retry ${retryCount}/${maxRetries} after ${delay}ms: ${lastError.message}`);
        }

        this.emit('retry', {
          requestId,
          attempt: retryCount,
          delay,
          error: lastError.message
        });

        await this.sleep(delay);
      }
    }

    this.recordMetrics({
      requestId,
      method: options.method,
      path: options.path,
      startTime,
      endTime: Date.now(),
      duration: Date.now() - startTime,
      status: 0,
      success: false,
      retryCount,
      fromCache: false
    });

    throw lastError!;
  }

  /**
   * Make the actual HTTP request
   */
  private async makeHttpRequest<T>(options: RequestOptions, requestId: string): Promise<ApiResponse<T>> {
    const url = `${this.baseUrl}${options.path}`;
    const method = options.method;
    const body = options.body ? JSON.stringify(options.body) : undefined;
    const timeout = options.timeout ?? this.timeout;

    // Prepare headers
    let headers: Record<string, string> = {
      'Content-Type': 'application/json',
      'User-Agent': 'SwappiQ-SDK-TS/1.0.0',
      'X-Request-ID': requestId,
      ...options.headers
    };

    // Sign request if authentication is required
    if (options.auth !== false && this.auth) {
      const signedRequest = await this.requestSigner.signRequest({
        method,
        path: options.path,
        body: body || '',
        timestamp: Date.now().toString(),
        signature: '',
        headers: {}
      });

      headers = { ...headers, ...signedRequest.headers };
    }

    if (this.debug) {
      console.log(`[${requestId}] ${method} ${url}`, {
        headers: this.sanitizeHeaders(headers),
        body: this.sanitizeBody(options.body)
      });
    }

    // Create abort controller for timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    try {
      const response = await fetch(url, {
        method,
        headers,
        body,
        signal: controller.signal
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        throw new HttpError(`HTTP ${response.status}: ${response.statusText}`, response.status);
      }

      const data = await response.json();
      
      // Validate response structure
      if (!this.isValidApiResponse(data)) {
        throw new Error('Invalid API response structure');
      }

      if (this.debug) {
        console.log(`[${requestId}] Response:`, this.sanitizeResponse(data));
      }

      return data as ApiResponse<T>;

    } catch (error) {
      clearTimeout(timeoutId);
      
      if (error instanceof Error && error.name === 'AbortError') {
        throw new TimeoutError(`Request timeout after ${timeout}ms`);
      }
      
      throw error;
    }
  }

  /**
   * Calculate exponential backoff delay with jitter
   */
  private calculateRetryDelay(attempt: number): number {
    const baseDelay = this.retryConfig.baseDelay;
    const backoffFactor = this.retryConfig.backoffFactor;
    const maxDelay = this.retryConfig.maxDelay;
    const jitter = this.retryConfig.jitter;

    let delay = Math.min(baseDelay * Math.pow(backoffFactor, attempt - 1), maxDelay);

    if (jitter) {
      // Add random jitter ±25%
      const jitterAmount = delay * 0.25;
      delay += (Math.random() - 0.5) * 2 * jitterAmount;
    }

    return Math.floor(delay);
  }

  /**
   * Check if error is retryable
   */
  private isRetryableError(error: any): boolean {
    if (error instanceof TimeoutError) return true;
    if (error instanceof HttpError) {
      return error.status >= 500 || error.status === 429; // Server errors and rate limiting
    }
    
    const errorMessage = error.message?.toLowerCase() || '';
    return this.retryConfig.retryableErrors.some(retryableError => 
      errorMessage.includes(retryableError.toLowerCase())
    );
  }

  /**
   * Validate API response structure
   */
  private isValidApiResponse(data: any): boolean {
    return (
      typeof data === 'object' &&
      data !== null &&
      typeof data.success === 'boolean' &&
      typeof data.timestamp === 'string' &&
      typeof data.requestId === 'string'
    );
  }

  /**
   * Create standardized API error
   */
  private createApiError(error: any, requestId: string): ApiError {
    if (error instanceof HttpError) {
      return {
        code: `HTTP_${error.status}`,
        message: error.message,
        details: { status: error.status, requestId },
        retryable: this.isRetryableError(error)
      };
    }

    if (error instanceof TimeoutError) {
      return {
        code: 'TIMEOUT',
        message: error.message,
        details: { requestId },
        retryable: true
      };
    }

    return {
      code: 'UNKNOWN_ERROR',
      message: error.message || 'An unknown error occurred',
      details: { requestId },
      retryable: false
    };
  }

  /**
   * Cache management
   */
  private getFromCache(path: string): CachedResponse | null {
    const cached = this.responseCache.get(path);
    if (!cached) return null;

    if (Date.now() - cached.timestamp > 30000) { // 30 second TTL
      this.responseCache.delete(path);
      return null;
    }

    return cached;
  }

  private cacheResponse(path: string, response: ApiResponse<any>): void {
    this.responseCache.set(path, {
      response,
      timestamp: Date.now()
    });
  }

  /**
   * Generate unique request ID
   */
  private generateRequestId(): string {
    return `req_${Date.now()}_${++this.requestCounter}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Get key for duplicate request detection
   */
  private getDuplicateKey(options: RequestOptions): string {
    const bodyHash = options.body ? JSON.stringify(options.body) : '';
    return `${options.method}:${options.path}:${bodyHash}`;
  }

  /**
   * Record request metrics
   */
  private recordMetrics(metrics: RequestMetrics): void {
    this.metrics.push(metrics);
    
    // Keep only last 1000 metrics
    if (this.metrics.length > 1000) {
      this.metrics = this.metrics.slice(-1000);
    }

    this.emit('metrics', metrics);
  }

  /**
   * Cleanup old data
   */
  private cleanup(): void {
    const now = Date.now();
    
    // Clear old cache entries
    for (const [key, cached] of this.responseCache.entries()) {
      if (now - cached.timestamp > 300000) { // 5 minutes
        this.responseCache.delete(key);
      }
    }

    // Clear old metrics
    this.metrics = this.metrics.filter(m => now - m.endTime < 3600000); // 1 hour
  }

  /**
   * Comprehensive sanitization for logging security
   */
  private sanitizeHeaders(headers: Record<string, string>): Record<string, string> {
    const sanitized = { ...headers };
    const sensitiveHeaders = [
      'authorization', 'x-api-key', 'x-api-secret', 'x-signature', 
      'x-passphrase', 'x-nonce', 'cookie', 'set-cookie'
    ];
    
    for (const [key, value] of Object.entries(sensitized)) {
      const lowerKey = key.toLowerCase();
      if (sensitiveHeaders.some(h => lowerKey.includes(h)) || 
          lowerKey.includes('secret') || 
          lowerKey.includes('private') ||
          lowerKey.includes('auth')) {
        sanitized[key] = '[REDACTED]';
      }
    }
    
    return sanitized;
  }

  private sanitizeBody(body: any): any {
    if (!body || typeof body !== 'object') return body;
    
    const sanitized = Array.isArray(body) ? [...body] : { ...body };
    const sensitiveFields = [
      'apiSecret', 'privateKey', 'password', 'signature', 'secret',
      'token', 'auth', 'credentials', 'passphrase', 'nonce'
    ];
    
    const sanitizeRecursive = (obj: any): any => {
      if (Array.isArray(obj)) {
        return obj.map(item => sanitizeRecursive(item));
      } else if (obj && typeof obj === 'object') {
        const result: any = {};
        for (const [key, value] of Object.entries(obj)) {
          const lowerKey = key.toLowerCase();
          if (sensitiveFields.some(field => lowerKey.includes(field))) {
            result[key] = '[REDACTED]';
          } else if (typeof value === 'object') {
            result[key] = sanitizeRecursive(value);
          } else {
            result[key] = value;
          }
        }
        return result;
      }
      return obj;
    };
    
    return sanitizeRecursive(sanitized);
  }

  private sanitizeResponse(response: any): any {
    if (!response || typeof response !== 'object') return response;
    
    const sanitized = { ...response };
    if (sanitized.data && typeof sanitized.data === 'object') {
      sanitized.data = this.sanitizeBody(sanitized.data);
    }
    
    // Also sanitize error details that might contain sensitive info
    if (sanitized.error && typeof sanitized.error === 'object') {
      sanitized.error = this.sanitizeBody(sanitized.error);
    }
    
    return sanitized;
  }

  /**
   * Utility method for delays
   */
  private sleep(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Get client statistics
   */
  getStats(): HttpClientStats {
    const recentMetrics = this.metrics.filter(m => Date.now() - m.endTime < 3600000); // Last hour
    const successfulRequests = recentMetrics.filter(m => m.success);
    const failedRequests = recentMetrics.filter(m => !m.success);

    return {
      totalRequests: recentMetrics.length,
      successfulRequests: successfulRequests.length,
      failedRequests: failedRequests.length,
      successRate: recentMetrics.length > 0 ? successfulRequests.length / recentMetrics.length : 0,
      averageResponseTime: recentMetrics.length > 0 
        ? recentMetrics.reduce((sum, m) => sum + m.duration, 0) / recentMetrics.length 
        : 0,
      cacheHitRate: recentMetrics.length > 0
        ? recentMetrics.filter(m => m.fromCache).length / recentMetrics.length
        : 0,
      activeCacheEntries: this.responseCache.size,
      activeRequests: this.activeRequests.size,
      rateLimiterStats: this.rateLimiter?.getStats()
    };
  }

  /**
   * Clear all caches and reset state
   */
  reset(): void {
    this.responseCache.clear();
    this.activeRequests.clear();
    this.metrics = [];
    this.requestCounter = 0;
  }

  /**
   * Graceful shutdown
   */
  async shutdown(): Promise<void> {
    // Wait for active requests to complete
    const activePromises = Array.from(this.activeRequests.values());
    if (activePromises.length > 0) {
      await Promise.allSettled(activePromises);
    }

    this.reset();
    this.removeAllListeners();
  }
}

// ========== SUPPORTING TYPES AND CLASSES ==========

interface CachedResponse {
  response: ApiResponse<any>;
  timestamp: number;
}

export interface HttpClientStats {
  totalRequests: number;
  successfulRequests: number;
  failedRequests: number;
  successRate: number;
  averageResponseTime: number;
  cacheHitRate: number;
  activeCacheEntries: number;
  activeRequests: number;
  rateLimiterStats?: any;
}

class HttpError extends Error {
  constructor(message: string, public readonly status: number) {
    super(message);
    this.name = 'HttpError';
  }
}

class TimeoutError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'TimeoutError';
  }
}

File: sdk/python/swappiq_sdk/http_client.py
"""
HTTP client with automatic retry and exponential backoff for SwappiQ Protocol
Author: SwappiQ Protocol
Description: Production-ready HTTP client with comprehensive error handling and retry logic
"""

import asyncio
import aiohttp
import json
import time
import logging
from typing import Any, Dict, List, Optional, Union
from dataclasses import dataclass, field
from urllib.parse import urljoin, urlencode
import random
import hashlib

from .types import (
    SDKConfig, ApiResponse, ApiError, AuthCredentials, 
    RetryConfig, RateLimitConfig
)
from .request_signer import RequestSigner
from .rate_limiter import RateLimiter

logger = logging.getLogger(__name__)

@dataclass
class RequestOptions:
    method: str
    path: str
    body: Optional[Dict[str, Any]] = None
    headers: Optional[Dict[str, str]] = None
    auth: bool = False
    timeout: Optional[float] = None
    retries: Optional[int] = None
    priority: str = "normal"  # low, normal, high

@dataclass
class RequestMetrics:
    request_id: str
    method: str
    path: str
    start_time: float
    end_time: float
    duration: float
    status: int
    success: bool
    retry_count: int
    from_cache: bool

class HttpError(Exception):
    """HTTP-specific error with status code"""
    def __init__(self, message: str, status: int):
        super().__init__(message)
        self.status = status

class TimeoutError(Exception):
    """Request timeout error"""
    pass

class HttpClient:
    """Production-grade HTTP client with enterprise features"""
    
    def __init__(self, config: SDKConfig):
        self.base_url = config.api_url.rstrip('/')
        self.auth = config.auth
        self.retry_config = config.retry_config
        self.rate_limit_config = config.rate_limit_config
        self.timeout = config.timeout
        self.debug = config.debug
        
        # Initialize components
        self.request_signer = RequestSigner(config.auth) if config.auth else None
        self.rate_limiter = RateLimiter(config.rate_limit_config) if config.rate_limit_config else None
        
        # State management
        self.response_cache: Dict[str, Any] = {}
        self.active_requests: Dict[str, asyncio.Task] = {}
        self.request_counter = 0
        self.metrics: List[RequestMetrics] = []
        
        # Session for connection pooling
        self.session: Optional[aiohttp.ClientSession] = None
        
    async def __aenter__(self):
        await self._create_session()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
        
    async def _create_session(self):
        """Create aiohttp session with optimal settings"""
        timeout = aiohttp.ClientTimeout(total=self.timeout)
        connector = aiohttp.TCPConnector(
            limit=100,  # Total connection pool size
            limit_per_host=30,  # Per-host connection limit
            keepalive_timeout=30,
            enable_cleanup_closed=True
        )
        
        self.session = aiohttp.ClientSession(
            timeout=timeout,
            connector=connector,
            headers={
                'User-Agent': 'SwappiQ-SDK-Python/1.0.0',
                'Content-Type': 'application/json'
            }
        )
        
    async def request(self, options: RequestOptions) -> ApiResponse:
        """Make a type-safe API request with automatic retry and rate limiting"""
        request_id = self._generate_request_id()
        start_time = time.time()
        
        try:
            # Rate limiting check
            if self.rate_limiter:
                await self.rate_limiter.acquire(options.priority)
                
            # Check for duplicate requests
            duplicate_key = self._get_duplicate_key(options)
            if duplicate_key in self.active_requests:
                if self.debug:
                    logger.debug(f"[{request_id}] Waiting for duplicate request: {duplicate_key}")
                return await self.active_requests[duplicate_key]
                
            # Check cache for GET requests
            if options.method == 'GET':
                cached = self._get_from_cache(options.path)
                if cached:
                    self._record_metrics(RequestMetrics(
                        request_id=request_id,
                        method=options.method,
                        path=options.path,
                        start_time=start_time,
                        end_time=time.time(),
                        duration=0,
                        status=200,
                        success=True,
                        retry_count=0,
                        from_cache=True
                    ))
                    return cached
                    
            # Create the request task
            request_task = asyncio.create_task(
                self._execute_request(options, request_id, start_time)
            )
            self.active_requests[duplicate_key] = request_task
            
            try:
                response = await request_task
                
                # Cache successful GET responses
                if options.method == 'GET' and response.success:
                    self._cache_response(options.path, response)
                    
                return response
            finally:
                self.active_requests.pop(duplicate_key, None)
                
        except Exception as error:
            self._record_metrics(RequestMetrics(
                request_id=request_id,
                method=options.method,
                path=options.path,
                start_time=start_time,
                end_time=time.time(),
                duration=time.time() - start_time,
                status=0,
                success=False,
                retry_count=0,
                from_cache=False
            ))
            raise self._create_api_error(error, request_id)
            
    async def _execute_request(self, options: RequestOptions, request_id: str, start_time: float) -> ApiResponse:
        """Execute request with retry logic"""
        last_error: Optional[Exception] = None
        retry_count = 0
        max_retries = options.retries if options.retries is not None else self.retry_config.max_attempts
        
        while retry_count <= max_retries:
            try:
                response = await self._make_http_request(options, request_id)
                
                self._record_metrics(RequestMetrics(
                    request_id=request_id,
                    method=options.method,
                    path=options.path,
                    start_time=start_time,
                    end_time=time.time(),
                    duration=time.time() - start_time,
                    status=200,
                    success=response.success,
                    retry_count=retry_count,
                    from_cache=False
                ))
                
                return response
                
            except Exception as error:
                last_error = error
                
                if retry_count == max_retries or not self._is_retryable_error(error):
                    break
                    
                retry_count += 1
                delay = self._calculate_retry_delay(retry_count)
                
                if self.debug:
                    logger.warning(f"[{request_id}] Retry {retry_count}/{max_retries} after {delay}s: {str(error)}")
                    
                await asyncio.sleep(delay)
                
        self._record_metrics(RequestMetrics(
            request_id=request_id,
            method=options.method,
            path=options.path,
            start_time=start_time,
            end_time=time.time(),
            duration=time.time() - start_time,
            status=0,
            success=False,
            retry_count=retry_count,
            from_cache=False
        ))
        
        raise last_error
        
    async def _make_http_request(self, options: RequestOptions, request_id: str) -> ApiResponse:
        """Make the actual HTTP request"""
        if not self.session:
            await self._create_session()
            
        url = urljoin(self.base_url, options.path)
        method = options.method.upper()
        
        # Prepare headers
        headers = {
            'X-Request-ID': request_id,
            **(options.headers or {})
        }
        
        # Prepare body
        data = None
        if options.body:
            data = json.dumps(options.body, separators=(',', ':'))
            
        # Sign request if authentication is required
        if options.auth and self.request_signer:
            signed_request = await self.request_signer.sign_request(
                method=method,
                path=options.path,
                body=data or '',
                timestamp=str(int(time.time() * 1000))
            )
            headers.update(signed_request.headers)
            
        if self.debug:
            logger.debug(f"[{request_id}] {method} {url}")
            
        timeout = aiohttp.ClientTimeout(total=options.timeout or self.timeout)
        
        try:
            async with self.session.request(
                method=method,
                url=url,
                data=data,
                headers=headers,
                timeout=timeout
            ) as response:
                
                if not response.ok:
                    raise HttpError(f"HTTP {response.status}: {response.reason}", response.status)
                    
                response_data = await response.json()
                
                # Validate response structure
                if not self._is_valid_api_response(response_data):
                    raise ValueError("Invalid API response structure")
                    
                if self.debug:
                    logger.debug(f"[{request_id}] Response: {self._sanitize_response(response_data)}")
                    
                return ApiResponse(**response_data)
                
        except asyncio.TimeoutError:
            raise TimeoutError(f"Request timeout after {options.timeout or self.timeout}s")
            
    def _calculate_retry_delay(self, attempt: int) -> float:
        """Calculate exponential backoff delay with jitter"""
        base_delay = self.retry_config.base_delay
        backoff_factor = self.retry_config.backoff_factor
        max_delay = self.retry_config.max_delay
        jitter = self.retry_config.jitter
        
        delay = min(base_delay * (backoff_factor ** (attempt - 1)), max_delay)
        
        if jitter:
            # Add random jitter ±25%
            jitter_amount = delay * 0.25
            delay += (random.random() - 0.5) * 2 * jitter_amount
            
        return max(0.1, delay)  # Minimum 100ms delay
        
    def _is_retryable_error(self, error: Exception) -> bool:
        """Check if error is retryable"""
        if isinstance(error, TimeoutError):
            return True
        if isinstance(error, HttpError):
            return error.status >= 500 or error.status == 429  # Server errors and rate limiting
            
        error_message = str(error).lower()
        return any(retryable_error.lower() in error_message 
                  for retryable_error in self.retry_config.retryable_errors)
                  
    def _is_valid_api_response(self, data: Any) -> bool:
        """Validate API response structure"""
        return (
            isinstance(data, dict) and
            'success' in data and
            isinstance(data['success'], bool) and
            'timestamp' in data and
            'request_id' in data
        )
        
    def _create_api_error(self, error: Exception, request_id: str) -> ApiError:
        """Create standardized API error"""
        if isinstance(error, HttpError):
            return ApiError(
                code=f"HTTP_{error.status}",
                message=str(error),
                details={'status': error.status, 'request_id': request_id},
                retryable=self._is_retryable_error(error)
            )
            
        if isinstance(error, TimeoutError):
            return ApiError(
                code='TIMEOUT',
                message=str(error),
                details={'request_id': request_id},
                retryable=True
            )
            
        return ApiError(
            code='UNKNOWN_ERROR',
            message=str(error),
            details={'request_id': request_id},
            retryable=False
        )
        
    def _get_from_cache(self, path: str) -> Optional[ApiResponse]:
        """Get response from cache"""
        cached = self.response_cache.get(path)
        if not cached:
            return None
            
        if time.time() - cached['timestamp'] > 30:  # 30 second TTL
            del self.response_cache[path]
            return None
            
        return cached['response']
        
    def _cache_response(self, path: str, response: ApiResponse) -> None:
        """Cache response"""
        self.response_cache[path] = {
            'response': response,
            'timestamp': time.time()
        }
        
    def _generate_request_id(self) -> str:
        """Generate unique request ID"""
        self.request_counter += 1
        return f"req_{int(time.time() * 1000)}_{self.request_counter}_{random.randint(1000, 9999)}"
        
    def _get_duplicate_key(self, options: RequestOptions) -> str:
        """Get key for duplicate request detection"""
        body_hash = ""
        if options.body:
            body_str = json.dumps(options.body, sort_keys=True, separators=(',', ':'))
            body_hash = hashlib.md5(body_str.encode()).hexdigest()
        return f"{options.method}:{options.path}:{body_hash}"
        
    def _record_metrics(self, metrics: RequestMetrics) -> None:
        """Record request metrics"""
        self.metrics.append(metrics)
        
        # Keep only last 1000 metrics
        if len(self.metrics) > 1000:
            self.metrics = self.metrics[-1000:]
            
    def _sanitize_response(self, response: Dict[str, Any]) -> Dict[str, Any]:
        """Sanitize response for logging"""
        sanitized = response.copy()
        if 'data' in sanitized and isinstance(sanitized['data'], dict):
            data = sanitized['data'].copy()
            sensitive_fields = ['api_secret', 'private_key', 'password', 'signature']
            for field in sensitive_fields:
                if field in data:
                    data[field] = '[REDACTED]'
            sanitized['data'] = data
        return sanitized
        
    def get_stats(self) -> Dict[str, Any]:
        """Get client statistics"""
        recent_metrics = [m for m in self.metrics if time.time() - m.end_time < 3600]  # Last hour
        successful_requests = [m for m in recent_metrics if m.success]
        failed_requests = [m for m in recent_metrics if not m.success]
        
        return {
            'total_requests': len(recent_metrics),
            'successful_requests': len(successful_requests),
            'failed_requests': len(failed_requests),
            'success_rate': len(successful_requests) / max(len(recent_metrics), 1),
            'average_response_time': sum(m.duration for m in recent_metrics) / max(len(recent_metrics), 1),
            'cache_hit_rate': len([m for m in recent_metrics if m.from_cache]) / max(len(recent_metrics), 1),
            'active_cache_entries': len(self.response_cache),
            'active_requests': len(self.active_requests),
            'rate_limiter_stats': self.rate_limiter.get_stats() if self.rate_limiter else None
        }
        
    def is_healthy(self) -> bool:
        """Check if client is healthy"""
        stats = self.get_stats()
        return (
            stats['success_rate'] > 0.95 and
            stats['average_response_time'] < 5.0 and  # Less than 5 seconds
            len(self.active_requests) < 50  # Not too many concurrent requests
        )
        
    async def close(self) -> None:
        """Clean up resources"""
        if self.session:
            await self.session.close()
            
        # Cancel active requests
        for task in self.active_requests.values():
            if not task.done():
                task.cancel()
                
        self.active_requests.clear()
        self.response_cache.clear()
        self.metrics.clear()